File: leetcode/2. problem/java/494.目标和.java
import java.util.Arrays;

/*
 * @lc app=leetcode.cn id=494 lang=java
 *
 * [494] 目标和
 */

// @lc code=start
class Solution {
    int count = 0;

    public int findTargetSumWays(int[] nums, int target) {
        // 回溯算法
        return extracted(nums, target);

        // 二维数组 动态规划 -- 01 背包
        // return extracted2(nums, target);

        // 一维数组 动态规划 -- 01 背包
        // return extracted3(nums, target);
    }

    private int extracted2(int[] nums, int target) {
        int sum = Arrays.stream(nums).sum();

        // 前置判断
        int diff = sum - target;
        if (diff < 0 || diff % 2 != 0)
            return 0;

        // 创建背包
        int n = nums.length, neg = diff / 2;
        int[][] dp = new int[n + 1][neg + 1];
        dp[0][0] = 1;
        for (int i = 1; i <= n; i++) {
            int num = nums[i - 1];
            for (int j = 0; j <= neg; j++) {
                dp[i][j] = dp[i - 1][j];
                if (j >= num) {
                    dp[i][j] += dp[i - 1][j - num];
                }
            }
        }
        return dp[n][neg];
    }

    private int extracted3(int[] nums, int target) {
        int sum = 0;
        for (int i = 0; i < nums.length; i++)
            sum += nums[i];
        if ((target + sum) % 2 != 0)
            return 0;
        int size = (target + sum) / 2;
        if (size < 0)
            size = -size;
        int[] dp = new int[size + 1];
        dp[0] = 1;
        for (int i = 0; i < nums.length; i++) {
            for (int j = size; j >= nums[i]; j--) {
                dp[j] += dp[j - nums[i]];
            }
        }
        return dp[size];
    }

    /**
     * 139/139 cases passed (548 ms)
     * Your runtime beats 14.47 % of java submissions
     * Your memory usage beats 55.26 % of java submissions (39.4 MB)
     * 
     * @param nums
     * @param target
     * @return
     */
    private int extracted(int[] nums, int target) {
        dfs(nums, target, 0, 0);
        return count;
    }

    private void dfs(int[] nums, int target, int index, int sum) {
        if (index == nums.length) {
            if (sum == target) {
                count++;
            }
        } else {
            dfs(nums, target, index + 1, sum + nums[index]);
            dfs(nums, target, index + 1, sum - nums[index]);
        }
    }
}
// @lc code=end

File: leetcode/2. problem/494.目标和.go
/*
 * @lc app=leetcode.cn id=494 lang=golang
 *
 * [494] 目标和
 */
package main

// @lc code=start
func findTargetSumWays(nums []int, target int) int {
	return newFunction22(nums, target)

}

// 140/140 cases passed (4 ms)
// Your runtime beats 62.33 % of golang submissions
// Your memory usage beats 23.12 % of golang submissions (6.4 MB)
func newFunction22(nums []int, target int) int {
	n := len(nums)
	sum := 0
	for _, v := range nums {
		sum += v
	}
	diff := sum - target
	if diff < 0 || diff%2 != 0 {
		return 0
	}
	neg := diff / 2
	dp := make([][]int, n+1)
	for i := 0; i <= n; i++ {
		dp[i] = make([]int, neg+1)
	}
	dp[0][0] = 1
	for i := 1; i <= n; i++ {
		val := nums[i-1]
		for j := 0; j <= neg; j++ {
			dp[i][j] = dp[i-1][j]
			if j >= val {
				dp[i][j] += dp[i-1][j-val]
			}
		}
	}
	return dp[n][neg]
}

// @lc code=end
